### ISSUE TITLE
Make SupervisorStrategy serializable

### ISSUE THREAD

**Issue by Aaronontheweb** [akkadotnet/akka.net#599]
One of the outstanding dependent issues we need to resolve before we can consider #545 resolved is a question @rogeralsing  raised in #581:

> The currently largest issue is the implementation of SupervisorStrategy, we use a delegate ('Decider') to deal with exception evaluation, this does not serialize.
>
> The only option I can think of is to replace this with a KVP -> exception type + directive and serialize that instead.

To which I added:

> For the `SupervisorStrategy` bit, we should still allow the use of delegates *but not as the default option*. There are some capabilities that supervisors need to do such as publish child failures to another subscriber (this gets used into the Akka.Remote internals for listening for transport errors.) Can't get rid of the delegate option entirely.
> 
> We should warn the end user that any supervisor strategy that uses a delegate is not serializable. Kind of like how we do with `DynamicProps` (which may no longer be needed, now that we have `IndirectActorProducer` support.)
> 
> For 80% of cases, using a `Decider` DSL that just does pattern matching around Exception types will get the job done and we can guarantee that that is serializable. We should make that the safe default and allow advanced users access to the delegate method, with the warning that you can't serialize this stuff over the wire.

Conclusion: we need a wire-friendly DSL for building `Deciders`. A `PatternMatch` equivalent for handling exceptions, issuing directives, and optionally being able to send a message to an `ActorRef` should be included in the design.

**Comment by rogeralsing** [comment]
What if we make the decider completely value based, but based on outcome, call some lifecycle events, where you could trigger various actions, instead of triggering them from within the decider itself. 
Would that work?

**Comment by Aaronontheweb** [comment]
I think so - depends on whether or not those actions need to be network-serializable, right?

**Comment by rogeralsing** [comment]
What if those actions belongs to the actor, not the decider?

decider Outcome -> actor.onXxxx(..)

**Comment by Aaronontheweb** [comment]
So here's a real-world example from one of our unit tests:

```csharp
 public void RemoteRouter_must_set_supplied_SupervisorStrategy()
        {
            var probe = CreateTestProbe(masterActorSystem);
            var escalator = new OneForOneStrategy(ex =>
            {
                probe.Ref.Tell(ex);
                return Directive.Escalate;
            });

            var router =
                masterActorSystem.ActorOf(
                    new RemoteRouterConfig(new RoundRobinPool(1, null, escalator, null),
                        new[] { new Address("akka.tcp", sysName, "localhost", port) })
                        .Props(Props.Empty), "blub3");

            router.Tell(new GetRoutees(), probe.Ref);
            // Need to be able to bind EventFilter to additional actor system (masterActorSystem in this case) before this code works
            // EventFilter.Exception<ActorKilledException>().ExpectOne(() => 
            probe.ExpectMsg<Routees>().Members.Head().Send(Kill.Instance, TestActor);
            //);
            probe.ExpectMsg<ActorKilledException>();
        }
```

In this case we're just pushing the data collected inside the `Decider` to a `TestProbe`, and there are some other cases like this one too.

One thought I had was we could accomplish this by adding something like this inside a Decider.DSL

    Decider.NewDecider()
    .Match<HeliosConnectionException>()
    	.Directive(Directive.Escalate)
    	.PublishException(SomeActorRef, 
    		ex => new WrappedClass(ex))
    .Match<IOException>()
    	.Directive(Directive.Restart)
    	.Publish(SomeOtherActorRef, 
			ex => ex) //don't wrap the Exception this time 
    .Default()
    	.Directive(Directive.Restart);

**Comment by Aaronontheweb** [comment]
err, meant to make both of those method calls read as `Publish`

Something like this would be serializable, even across network deployments - so network-deployed routers could still report exceptions back to the machine that did the original deploying.

**Comment by rogeralsing** [comment]
Is this how it is in the Scala code also?

I'm looking at http://doc.akka.io/api/akka/2.2.0/index.html#akka.actor.SupervisorStrategy$

```
def makeDecider(flat: Iterable[(Class[_ <: Throwable], Directive)]): Decider

Decider builder for Iterables of cause-directive pairs, e.g. a map obtained from configuration; will sort the pairs so that the most specific type is checked before all its subtypes, allowing carving out subtrees of the Throwable hierarchy.
```

Looks like the decider in scala is KVP of exception+directive, no?

**Comment by rogeralsing** [comment]
So what if we make `Decider` a base class, and then add a subtype that can decide from a func.
And another subtype that uses a KVP of exception+directive.
And we only let the KVP decider be serializable, then we can still pass around that between nodes.

**Comment by rogeralsing** [comment]
cc @akkadotnet/developers are we all OK with this design? see my comment above.
One serializable decider and one that is not..

**Comment by Aaronontheweb** [comment]
I approve of this design - Delegate based Deciders for folks who need stuff that can't serialize anyway, but DSL-based deciders using the design you mentioned by default.

**Comment by rogeralsing** [comment]
Working on this one

**Comment by HCanber** [comment]
Somewhat late, but yes, it sounds ok with a few tweaks. 
Create an `IDecider` interface and use that instead of a base class.
Make `Decider` a static class and put the DSL in that.
`DelegateDecider` and `SerializableDecider` (or whatever we call them) implements `IDecider` and only hold the values, no DSLs here.
If they share functionality create a new base class, but let the framework depend on `IDecider` and not on the base class.

**Comment by rogeralsing** [comment]
@HCanber that is how I did it in https://github.com/akkadotnet/akka.net/pull/665

**Comment by rogeralsing** [comment]
I did however keep the old overloads using Func also for backwards compat

**Comment by Aaronontheweb** [comment]
Resolved as of #665
